File: src/main/java/miragefairy2019/mod/modules/fairy/registry/FairySelector.java
package miragefairy2019.mod.modules.fairy.registry;

import java.util.HashSet;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Predicate;

import miragefairy2019.mod.api.fairy.IFairySelector;
import mirrg.boron.util.suppliterator.ISuppliterator;
import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;

public class FairySelector implements IFairySelector
{

	private FairyRelationRegistry fairyRelationRegistry;

	public FairySelector(FairyRelationRegistry fairyRelationRegistry)
	{
		this.fairyRelationRegistry = fairyRelationRegistry;
	}

	//

	private Set<ItemStack> itemStacks = new HashSet<>();

	@Override
	public IFairySelector add(ItemStack... itemStacks)
	{
		ISuppliterator.ofObjArray(itemStacks)
			.forEach(this.itemStacks::add);
		return this;
	}

	private Set<IBlockState> blockStates = new HashSet<>();

	@Override
	public IFairySelector add(IBlockState... blockStates)
	{
		ISuppliterator.ofObjArray(blockStates)
			.forEach(this.blockStates::add);
		return this;
	}

	//

	@Override
	public ISuppliterator<ResourceLocation> select()
	{
		return ISuppliterator.ofIterable(fairyRelationRegistry.map.entrySet())
			.filter(e -> {
				FairyRelation fairyRelation = e.getValue();

				for (Predicate<ItemStack> predicate : fairyRelation.predicatesItemStack) {
					for (ItemStack itemStack : itemStacks) {
						if (predicate.test(itemStack)) return true;
					}
				}

				for (Predicate<IBlockState> predicate : fairyRelation.predicatesBlockState) {
					for (IBlockState blockState : blockStates) {
						if (predicate.test(blockState)) return true;
					}
				}

				return false;
			})
			.map(Entry::getKey);
	}

}

File: src/main/java/miragefairy2019/mod/fairyrelation/FairySelector.kt
package miragefairy2019.mod.fairyrelation

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.Entity
import net.minecraft.init.Blocks
import net.minecraft.init.Items
import net.minecraft.item.Item
import net.minecraft.item.ItemStack

class FairySelector {
    private val entities = mutableSetOf<Entity>()
    private val blocks = mutableSetOf<Block>()
    private val blockStates = mutableSetOf<IBlockState>()
    private val items = mutableSetOf<Item>()
    private val itemStacks = mutableSetOf<ItemStack>()

    fun entity(entity: Entity): FairySelector {
        entities += entity
        return this
    }

    fun blockState(blockState: IBlockState): FairySelector {
        blockStates += blockState
        blocks += blockState.block
        val item = Item.getItemFromBlock(blockState.block)
        if (item != Items.AIR) items += item
        return this
    }

    fun itemStack(itemStack: ItemStack): FairySelector {
        itemStacks += itemStack
        items += itemStack.item
        val block = Block.getBlockFromItem(itemStack.item)
        if (block != Blocks.AIR) blocks += block
        return this
    }

    fun allMatch(): List<FairyRelationEntry<*>> = listOf(
        FairyRelationRegistries.entity.entries.filter { entities.any { entity -> it.key(entity) } },
        FairyRelationRegistries.block.entries.filter { it.key in blocks },
        FairyRelationRegistries.blockState.entries.filter { it.key in blockStates },
        FairyRelationRegistries.item.entries.filter { it.key in items },
        FairyRelationRegistries.itemStack.entries.filter { itemStacks.any { itemStack -> it.key(itemStack) } },
        FairyRelationRegistries.ingredient.entries.filter { itemStacks.any { itemStack -> it.key.test(itemStack) } }
    ).flatten()
}

/** [FairyRelationEntry.relevance]が1に満たないものを除外します。 */
val <T : FairyRelationEntry<*>> List<T>.withoutPartiallyMatch get() = filter { it.relevance >= 1.0 }

/** 最も[FairyRelationEntry.relevance]の大きなものの集合を返します。 */
val <T : FairyRelationEntry<*>> List<T>.primaries: List<T>
    get() {
        if (isEmpty()) return listOf() // 空の場合は空を返す
        val maxRelevance = map { it.relevance }.max()!!
        return filter { it.relevance == maxRelevance }
    }
